Support multi instance endorsement (#434)
ietf-rats-wg/draft-ietf-rats-corim: draft-ietf-rats-corim.md

The types defined for a group identified are UUID and variable-length opaque byte string ({{sec-common-tagged-bytes}}).

Wild-card binding can be used within the `group` field, see {{sec-comid-wc}}.

~~~ cddl
{::include cddl/group-id-type-choice.cddl}
~~~

##### Environment Wild-card Binding {#sec-comid-wc}

The `tagged-var-bind` and `tagged-var-ref` values are used for constrained wildcard processing.

~~~ cddl
{::include cddl/wc-match-type-choice.cddl}
~~~

An `environment-map` without an `instance` field used in a condition will match against an ACS-ECT with any `instance` field.

> When the ACS contains evidence from multiple Attesters, and that evidence uses the `instance` field to indicate which Attester created the evidence, these values are used to indicate the relationships between environments.

Author: ```suggestion
When the ACS contains evidence from multiple Target Environments, and that evidence uses the `instance` field to indicate which Target Environment created the evidence, these values are used to indicate the relationships between environments.
```

The types defined for a group identified are UUID and variable-length opaque byte string ({{sec-common-tagged-bytes}}).

Wild-card binding can be used within the `group` field, see {{sec-comid-wc}}.

~~~ cddl
{::include cddl/group-id-type-choice.cddl}
~~~

##### Environment Wild-card Binding {#sec-comid-wc}

The `tagged-variable` value is used for constrained wildcard processing.

~~~ cddl
{::include cddl/wc-match-type-choice.cddl}
~~~

An `environment-map` without an `instance` field used in a condition will match against an ACS-ECT with any `instance` field.
When the ACS contains evidence from multiple Target Environments, and that evidence uses the `instance` field to indicate which Target Environment the evidence relates to, the `tagged-variable` values are used to indicate the relationships between `environment-map`s.
These values behave in a similar way when used in the `group` field.

In a condition, the `tagged-variable` is used to indicate that two or more `environment-map`s must have the same `instance` value.

> In an addition, the `tagged-variable` value is used to indicate that the `environment-map`.`instance` field must be copied from the ACS-ECT which matched a condition within the same triple.

Author: ```suggestion
In an addition, the `tagged-variable` value is used to indicate that the `environment-map`.`instance` field must be copied from the ACS-ECT which matched a condition within the same triple, to the ECT added as a result of the addition.
```

#### Processing Endorsements {#sec-process-end}

Endorsed Value Triple and Conditional Endorsement Triple share the same internal representation.
Both types of triple are transformed into an internal representation based on `ev`, but the CBOR encoding of Endorsed Value Triple can only represent a subset of the CBOR encoding of Conditional Endorsement Triple.

> The Endorsed Values Triple contains a single `environment-map`, which means it can only match against one environment in the ACS; it can only add a single Endorsement ECT; and it uses the same enviromnent for `condition` and `addition`.

Reviewer: ```suggestion
The Endorsed Values Triple contains a single `environment-map`, which means it can only match against one environment in the ACS; it can only add Endorsements for ECT; and pertaining to one Environment
```

Author: Added similar fix. Thanks.

> After transformation into an `ev` entry, the processing steps of both triples are the same, as described below.

Author: The text is talking about what happens after the triples are transformed into `ev` form. What I'm saying is that after this transform the processing is the same.

For example, an endorsed values triple is transformed into an `ev` with:
`condition.environment = environment`
`condition.measurement = <empty>`
`addition.environment = environment`
`addition.measurement = measurement`

A conditional endorsement triple can create exactly the same `ev`.
The results of processing this `ev` will be the same regardless of whether it came from an Endorsed Values Triple or a Conditional Endorsement Triple.

Each `ev` entry is processed independently of other `ev`s.

The Verifier SHALL find all sets of ACS-ECTs which match all the `ev`.`condition`s, where the ACS-ECT `cmtype` contains either `evidence`, `reference-values`, or `endorsements`.
For each match ({{sec-match-condition-ect}}), all the `ev`.`addition` ECTs are added to the ACS.

Note that some condition values are able to match against multiple ACS-ECTs, or sets of ACS-ECTs.
If there are multiple matches then each match is processed independently from the others.

##### Copying instance and/or group fields from a condition {#sec-comid-instance-group-copy}

A CoRIM author may need to create a conditional endorsement which applies to all all measurements which have the same `class` field within their `environment-map`, regardless of their `instance` and/or `group` fields.
The `slot-bind-type` and `slot-ref-type` options in these fields can be used to achieve this.

> In the simplest case, the CoRIM sets the `enviroment-map`.`instance` field of a `stateful-environment-record` within the triple to hold an `slot-bind-type` and the `enviroment-map`.`instance` field of an `endorsed-triple-record` to hold an `slot-ref-type`.

Reviewer: ```suggestion
In the simplest case, the CoRIM author sets the `environment-map`.`instance` field of a `stateful-environment-record` within the triple to hold an `slot-bind-type` and the `environment-map`.`instance` field of an `endorsed-triple-record` to hold an `slot-ref-type`.
```

Both instance types contain an integer name for the slot variable which is bound by the `slot-bind-type` and referenced by the `slot-ref-type`.

Within each triple, each slot variable can only be bound to one value, so if there are multiple `environment-map`s using `slot-bind-type` then subsequent uses introduce a constraint that the value to bind must be equal to the value already bound to the slot.

> After successfully matching a `stateful-environment-record` containing an `slot-bind-type` against an ACS entry, the verifier SHALL copy the instance value from that ACS entry to the corresponding slot variable.

Reviewer: It is not clear what is copied from which source to which destination?

Can you please add a clarity here...

If the matching ACS entry does not include an instance then the slot variable is marked as bound to the empty value.

> When adding a conditional endorsement whose `enviroment-map`.`instance` field is an `slot-ref-type` to the ACS, and that variable is bound to a non-empty value, then the verifier SHALL set the `enviroment-map`.`instance` field from the contents of the corresponding slot variable.

Reviewer: `enviroment-map`.`instance` field from the contents of the corresponding slot variable to where ??

In the simplest case, the CoRIM sets the `enviroment-map`.`instance` field of a `stateful-environment-record` within the triple to hold an `instance-slot-store-type` and the `enviroment-map`.`instance` field of an `endorsed-triple-record` to hold an `instance-slot-use-type`.
Both instance types contain an integer name for the variable which is set by the `instance-slot-store-type` and used by the `instance-slot-use-type`.

Within each triple, each variable can only be set once, so if there are multiple `environment-map`s using `instance-slot-store-type` then they must each have a different integer name.

After successfully matching a `stateful-environment-record` containing an `instance-slot-store-type` against an ACS entry, the verifier SHALL copy the instance value from that ACS entry to the corresponding slot variable.
If the matching ACS entry does not include an instance then the slot variable is marked as initialised, but empty.

When adding a conditional endorsement whose `enviroment-map`.`instance` field is an `instance-slot-use-type` to the ACS, the verifier SHALL set the `enviroment-map`.`instance` field from the corresponding slot variable.

> If the corresponding slot variable is initialised but empty then the verifier SHALL NOT add an `instance` field.

Reviewer 2: ```suggestion
If the corresponding slot variable is initialized but empty then the verifier SHALL NOT add an `instance` field.
```

The `instance-slot-bind-type` and `instance-slot-ref-type` options in the `instance` field can be used to achieve this.

In the simplest case, the CoRIM sets the `enviroment-map`.`instance` field of a `stateful-environment-record` within the triple to hold an `instance-slot-bind-type` and the `enviroment-map`.`instance` field of an `endorsed-triple-record` to hold an `instance-slot-ref-type`.
Both instance types contain an integer name for the slot variable which is bound by the `instance-slot-bind-type` and referenced by the `instance-slot-ref-type`.

Within each triple, each slot variable can only be bound to one value, so if there are multiple `environment-map`s using `instance-slot-bind-type` then subsequent uses introduce a constraint that the value to bind must be equal to the value already bound to the slot.

After successfully matching a `stateful-environment-record` containing an `instance-slot-bind-type` against an ACS entry, the verifier SHALL copy the instance value from that ACS entry to the corresponding slot variable.
If the matching ACS entry does not include an instance then the slot variable is marked as bound to the empty value.

> When adding a conditional endorsement whose `enviroment-map`.`instance` field is an `instance-slot-ref-type` to the ACS, the verifier SHALL set the `enviroment-map`.`instance` field from the contents of the corresponding slot variable.

Author: ```suggestion
When adding a conditional endorsement whose `enviroment-map`.`instance` field is an `instance-slot-ref-type` to the ACS, and that variable is bound to a non-empty value, then the verifier SHALL set the `enviroment-map`.`instance` field from the contents of the corresponding slot variable.
```

If the corresponding slot variable is bound to the empty value then the verifier SHALL NOT add an `instance` field.
If the corresponding slot variable is not bound then the triple is invalid - the verifier SHALL NOT add the conditional endorsement to the ACS.

If a conditional endorsement containing stateful environments which use `instance-slot-bind-type` matches against multiple ECTs, then each match is processed independently of the others.

> Each match uses its own slot bindings, and  each match adds a separate endorsement ECT to the ACS.

Author: ```suggestion
Each match uses its own slot bindings, and each match adds a separate endorsement ECT to the ACS.
```

For each match ({{sec-match-condition-ect}}), all the `ev`.`addition` ECTs are added to the ACS.

Note that some condition values can match against multiple ACS-ECTs, or sets of ACS-ECTs.
If there are multiple matches then each match is processed independently from the others.

##### Copying instance and/or group fields from a condition {#sec-comid-instance-group-copy}

A CoRIM author may need to create a conditional endorsement which applies to all measurements which have the same `class` field within their `environment-map`, regardless of their `instance` and/or `group` fields.
The `tagged-var-bind` and `tagged-var-ref` options in these fields can be used to achieve this.

In the simplest case, within a CoRIM, the `enviroment-map`.`instance` field of an `ev` condition holds a `tagged-var-bind` and the `enviroment-map`.`instance` field of an addition within that `ev` holds a `tagged-var-ref`.
Both instance types contain an integer name for the bound variable, which is bound by the `tagged-var-bind` and referenced by the `tagged-var-ref`.

> Within each triple, each bound variable can only be bound to one value, so if there are multiple `environment-map`s using `tagged-var-bind` then subsequent uses introduce a constraint that the value to bind must be equal to the value already bound to the slot.

Author: ```suggestion
Within each triple, each bound variable can only be bound to one value, so if there are multiple `environment-map`s using `tagged-var-bind` then subsequent uses introduce a constraint that the value to bind must be equal to the value already bound to the variable.
```

> Within each triple, each bound variable can only be bound to one value, so if there are multiple `environment-map`s using `tagged-var-bind` then subsequent uses introduce a constraint that the value to bind must be equal to the value already bound to the variable.

Author: TODO: Find a way to describe this without using bound variable terminology

> After successfully matching the `class` field within a `stateful-environment-record` whose `instance` field is a `tagged-var-bind` against an ACS entry, the Verifier SHALL copy the `instance` value from that ACS entry to the corresponding bound variable.

Reviewer: copy the `instance` value from that ACS entry to the corresponding bound variable?

Did you mean the ACS entry from Evidence?

> After successfully matching the `class` field within a `stateful-environment-record` whose `instance` field is a `tagged-var-bind` against an ACS entry, the Verifier SHALL copy the `instance` value from that ACS entry to the corresponding bound variable.

Author: ```suggestion
After successfully matching the `class` field within a `ev` condition whose `instance` field is a `tagged-var-bind` against an ACS entry, the Verifier SHALL copy the `instance` value from that ACS entry to the corresponding bound variable.
```

If the matching ACS entry does not include an instance then the bound variable is marked as bound to the empty value.

When adding an `ev` addition ECT whose `enviroment-map`.`instance` field is a `tagged-var-ref` to the ACS, if the variable is bound to a non-empty value, then the Verifier SHALL copy the value of the corresponding bound variable to the `enviroment-map`.`instance` field of the new ACS-ECT.
If the corresponding bound variable is bound to the empty value then the Verifier SHALL NOT add an `instance` field.

> If the corresponding bound variable is not bound then the triple is invalid - the Verifier SHALL NOT add the addition ECT to the ACS.

Reviewer 2: ```suggestion
If the corresponding bound variable is not bound then the triple is invalid—the Verifier SHALL NOT add the addition ECT to the ACS.
```

> If the corresponding bound variable is not bound then the triple is invalid - the Verifier SHALL NOT add the conditional endorsement to the ACS.

Author: ```suggestion
If the corresponding bound variable is not bound then the triple is invalid - the Verifier SHALL NOT add the addition ECT to the ACS.
```

#### Processing Endorsements {#sec-process-end}

Endorsed Value Triple and Conditional Endorsement Triple share the same internal representation.

> Both types of triple are transformed into an internal representation based on `ev`, but the CBOR encoding of Endorsed Value Triple can only represent a subset of the CBOR encoding of Conditional Endorsement Triple.

Reviewer: ```suggestion
Both types of triple are transformed into an internal representation based on `ev`, but the CBOR encoding of Endorsed Values Triple can only represent a subset of the CBOR encoding of Conditional Endorsement Triple.
```

After successfully matching an `ev` condition ECT whose `instance` field is a `tagged-variable` against an ACS entry, the Verifier SHALL bind the `instance` value to the `instance` value of the ACS-ECT which matched against that condition ECT.
If the matching ACS-ECT does not include an instance then the bound `instance` value is the empty value.

An `ev` addition ECT whose `enviroment-map`.`instance` field is a `tagged-variable` references the bound value.
When creating the ECT which will be added to the ACS, if the referenced `tagged-variable` is bound to a non-empty value, then the Verifier SHALL copy that bound value to the `enviroment-map`.`instance` field of the ACS-ECT created by the `ev` addition.
If the referenced variable is bound to the empty value then the Verifier SHALL NOT add an `instance` field.
If the referenced variable is not bound then the triple is invalid - the Verifier SHALL NOT add the addition ECT to the ACS.

> If a conditional endorsement containing `ev` conditions which use `tagged-variables` matches against multiple ECTs, then each match is processed independently of the others.

Reviewer: Such muddling is not helping, `addition` then again followed by `condition` again

Otherwise, do not add the `addition` ECT to the ACS.

#### Copying instance field from a condition {#sec-comid-instance-copy}

> A CoRIM author may need to create a conditional endorsement which applies to all devices of a particular type, regardless of their instance field.

Reviewer 2: ```suggestion
A CoRIM author may need to create a conditional endorsement which applies to all devices of a particular type, regardless of their instance field.
```
We don't talk about devices and their types. We only talk about environments and their class/group/instance. What does this mean?

Author: ```suggestion
A CoRIM author may need to create a conditional endorsement which applies to all measurements which have the same `global` and/or `class` field within their `environment-map`, regardless of their `instance` field.
```

Reviewer 3: We didn't define `global`. Where is this coming from?

Author: Text has been changed, this comment is included in those changes.

> If the first `stateless-environment` in the `conditions` field does not contain an `instance` codepoint then it will match against any environment with the same `class`.

Reviewer 2: ```suggestion
If the first `stateless-environment` in the `conditions` field does not contain an `instance` codepoint, then it will match against any environment with the same `class`.
```
This is roughly already the case, modulo `class` being `class` and/or `group`.

Reviewer 3: > `stateless-environment`

This isn't in the cddl. Possibly it should say "environment"?

In general, matching applies to all the environment fields. If there is a class, then it will match the class even if there is also an instance or group. The interesting case is if there isn't a class but you want the addition to apply to a class of ECTs.

> If the endorsement `environment-map` does not contain an `instance` field then the endorsements will be added without an instance field.

Reviewer 2: ```suggestion
If the endorsement `environment-map` does not contain an `instance` field, then the endorsements will be added without an instance field.
```

Reviewer 2: This doesn't match my understanding. You add endorsements to the ECTs that are selected by the condition, which can also have an existing instance field. Your words suggest that there'd be an ECT without an instance field with these endorsements.

When you discuss "the endorsement" you're talking about not the endorsement triples. You're talking about conditional endorsement (series) triples' non-condition component. It's a bit too imprecise.

Author: The text 3 lines ago said that this applies to conditional endorsements. A conditional endorsement can add the endorsement to any `environment-map` + `mkey` location.

Reviewer 3: > this applies to conditional endorsements

I don't think it applies to series endorsements since `conditional-series-record`.`addition` is only `measurement-map`. The ECT(s) updated are those  that satisfy the `condition` and `selection` criteria.

Only `conditional-endorsement-triple-record` can create a completely new ECT or update a different ECT from the ones that satisfied the `condition`.

This may cause problems if the ACS contains evidence from multiple target environments of the same type.

> The endorsement `enviromnent-map`/`instance` can be tagged to indicate that it is an `instance-copy`, with an integer argument.

Reviewer 2: ```suggestion
The endorsement's `environment-map`.`instance` can be tagged to indicate that it is an `instance-copy`, with an integer argument.
```

Author: Change added.

This type instructs the verifier to copy the `instance` field from the referenced condition to the endorsement.

> The integer argument is used to select which instance field to copy.

Reviewer 2: This is a different paragraph, so you've lost the context about what this argument is. Arguments are for function calls. Fields are for objects/maps. I'm not sure what you're talking about.

Instance-copy-type references "which instance field" to copy, which doesn't make sense given that instance is a field of environment, and there's no multiplicity to select amongst, no?

The following description is now imposing additional semantics to the appraisal algorithm that the ACS's order and/or the condition matching order is relevant to the interpretation of the triple, which is overconstrained and underspecified in the appraisal algorithm description section.

Author: This text has been rewritten to use named variables instead of positional references.

If the referenced variable is not bound then the triple is invalid - the Verifier SHALL NOT add the addition ECT to the ACS.

If a conditional endorsement containing `ev` conditions which use `tagged-variables` matches against multiple ECTs, then each match is processed independently of the others.
Each match uses its own variable bindings, and each match adds a separate addition ECT to the ACS.

The `tagged-variable` type in the `group` field work in the same way as in `instance`.
The `var-number`s for instance and group variables are separate, instance `var-number` 0 and group `var-number` 0 may be bound to different values.

##### Example algorithm for processing Tagged Variable

This section describes an algorithm for processing an `ev` containing two or more conditions which include `tagged-variable`s.

{:dmt-enum: counter="dmt1" style="format Step %d."}

{: dmt-enum}
* For each of the conditions, find the set of ACS-ECTs which match the fields within that condition which are not `tagged-variable`s.

* Iterate through the set of AES-ECTs which match the first condition. There may be multiple matches with different `instance` values.

> Any `tagged-variable`s within the first condition will be bound to values from the matching `instance` AES-ECTs.

Reviewer: ```suggestion
Any `tagged-variable`s within the first condition will be bound to values from the matching `instance` ACS-ECTs.
```

> This section describes an algorithm for processing an `ev` containing two or more conditions which include `tagged-variable`s.
> 
> {:dmt-enum: counter="dmt1" style="format Step %d."}
> 
> {: dmt-enum}
> * For each of the conditions, find the set of ACS-ECTs which match the fields within that condition which are not `tagged-variable`s.
> 
> * Iterate through the set of ACS-ECTs which match the first condition. There may be multiple matches with different `instance` values.
> 
> Any `tagged-variable`s within the first condition will be bound to values from the matching `instance` ACS-ECTs.
> 
> {:dmt4-enum: counter="dmt4" style="format %i"}
> {: dmt4-enum}
> * Find ACS-ECTs which match the second condition, in the context of the bound `tagged-variables`
> 
> * Find ACS-ECTs which match the third and subsequent conditions (if there are any).
> 
> * If any of the conditions has no matches for any addition ECT then discard that partial set.
> 
> * If any of the conditions has multiple matches for an ECT then create a partial set for each output.
> 
> {: dmt-enum}
> * Create a result set for each set of AES-ECTs which match all conditions, add that set to the list of successful matches.
> 
> * Continue processing the triple, processing it separately for each result set.
> For example, if the triple is a Conditional Endorsement triple which uses `tagged-variable` within the addition `instance` field then a separate AES-ECT will be added for each result set.

Reviewer: ```suggestion
[^tracked-at]  https://github.com/ietf-rats-wg/draft-ietf-rats-corim/issues/458
```

If a field is not present in the condition ECT `environment-map` then the presence of, and value of, the corresponding ACS entry field SHALL NOT affect whether the environments match.

#### Instance slot store and use fields

These fields are used to ensure that multiple conditions, for example within a conditional endorsement, match against the same instance value.
They are also used to set the instance value in a conditional endorsement ACS entry to match the matched ECTs.

The verifier maintains a small array of instance slots, each slot is identified using a non-negative integer.
Before processing each triple, all slots SHALL be set to the unused state.

If the condition ECT `environment-map` contains an `instance` field of type `instance-slot-store-type` then it matches against any instance value.
The instance field in the ACS entry is copied into the instance slot with the relevant `instance-slot-number`.

> If the condition ECT `environment-map` contains an `instance` field of type `instance-slot-use-type` and that slot is unused then the environments do not match.

Reviewer 2: ```suggestion
If the condition ECT `environment-map` contains an `instance` field of type `instance-slot-reference-type` and that slot is unused then the environments do not match.
```

> If the condition ECT `environment-map` contains an `instance` field of type `instance-slot-use-type` and it is not binary identical to the value in that slot then the environments do not match.

Reviewer 2: ```suggestion
If the condition ECT `environment-map` contains an `instance` field of type `instance-slot-reference-type` and it is not binary identical to the value in that slot then the environments do not match.
```

> If a triple contains two or more condition ECTs which store values into the same instance slot then it is invalid and the environments do not match.

Reviewer 2: Delete.

|     562 | `bytes`             | tagged-pkix-asn1der-cert-type, see {{sec-crypto-keys}}        | {{&SELF}} |
|     563 | `tagged-masked-raw-value` | tagged-masked-raw-value, see {{sec-comid-raw-value-types}} | {{&SELF}} |
|     564 | `array`             | tagged-int-range, see {{sec-comid-int-range}}                   | {{&SELF}} |

> |     565 | `instance-slot-store` | instance-slot-store, see {{sec-comid-instance-copy}}    | {{&SELF}} |

Reviewer 2: ```suggestion
|     565 | `instance-slot-store` | instance-slot-bind, see {{sec-comid-instance-copy}}    | {{&SELF}} |
```

> |     566 | `instance-slot-use` | instance-slot-use, see {{sec-comid-instance-copy}}         | {{&SELF}} |

Reviewer 2: ```suggestion
|     566 | `instance-slot-use` | instance-slot-reference, see {{sec-comid-instance-copy}}         | {{&SELF}} |
```
